### PR TITLE
Clean up AstClause class.

### PR DESCRIPTION
**Pull request by azreika**
This pull request heavily refactors the AstClause class and its uses, finishing up the job started by PR #1279. The aim is to reduce the class structure so that it only contains what is necessary to describe a clause, as described in issue #1255. All unnecessary methods/structures/etc. have been moved to utils. 

Main changes include:
* **Removed the `FixedExecutionPlan` flag in AstClause, and all associated methods.** Although set in certain areas, the flag had no actual consequences on evaluation, and did not serve a purpose. 
* **Removed the `.strict` keyword from the parser.** The qualifier, added to the end of clauses, did not have any semantic consequences in the code, and did not affect any part of the Datalog engine. It was therefore removed to reduce bloat. Tests were updated to match this change.
* **Removed the `generated` flag in AstClause.** The assumed purpose of the `generated` flag was to avoid generating warnings and errors for clauses generated in the parser. In practice, the `generated` flag was only set when disjunctions were expanded in the parser into separate conjunction-only clauses. The consequence was that certain warnings were muted for disjunctive clauses, which is the bug behind issues #855 and #829.
* **Fixed muted warnings for disjunctive clauses.** Reasoning explained in previous point. Note that unused-variable warnings are now given for some clauses with disjunction even if the variable is used in one branch. E.g. `a(x), b(x,y) :- c(x), y = 0.` will give a warning for unused variable `y`, since it is not used for the `a(x)` head branch. This behaviour can be changed/specialised/etc. if `AstDisjunction` is added (see issue #1289).
* **Moved `clauseNum` from AstClause into a utils function.** The `clauseNum` variable is used only to uniquely identify clauses in relations for the provenance transformer. NOTE: in the old and new code there is an implicit assumption that the ID does not change between the provenance transformer and the translator to RAM. This seems a bit iffy, but is outside the scope of this PR.
* **Moved `reorderAtoms` into a utils function.** Note that the new structure of the clause class means that the reordering can no longer be in-place. Instead, a new clause with the new order is returned. This has a few cascading changes.
* **Fixed a bug where singleton variables for some variables were not being reported.** In particular, any variables starting with an underscore, even if named, were ignored. This is an unreported bug.
* **Changed variables of the form _NAME to indicate singleton variables.** If a user uses the same singleton-marked variable (e.g. `_Anybody`) in the same clause, then a warning is issued (the code will still behave as before).

Contributes towards the fix for issue #1255. Fixes issue #855 and issue #829.

### CODE REVIEW

**Comment by azreika** [comment]
In a clause like:

`bad(Dog) :- bites(Dog, _Anybody).`

do we:

* Give a single-use warning for `_Anybody`, and treat it like a named variable, or
* Not give a warning, and transform it into an unnamed variable in the back?

The system before this PR did not give a warning, but also did not treat it like an unnamed variable.

I am leaning towards the second, but then with a clause like:

`a(X) :- b(X, _X), c(_X).`

Do we want the `_X` to mean different things, or the same (i.e. bound to each other)?

This PR just gives a single-use warning for `_Anybody` and treats it like any other named variable.

File: src/AstClause.h [hunk 1]
```diff
@@ -144,15 +144,14 @@ class AstExecutionPlan : public AstNode {
  *       clauses, such as rules, queries and facts. This solution was to quickly
```
Comment (b-scholz) [inline src/AstClause.h]: Can you update the remarks; I feel they are misplaced with the latest refactoring.

File: src/AstClause.h [hunk 2]
```diff
@@ -183,45 +182,6 @@ class AstClause : public AstNode {
         return toPtrVector(bodyLiterals);
     }
 
-    /**
-     * Re-orders atoms to be in the given order.
-     * Remaining body literals remain in the same order.
-     **/
-    // TODO (b-scholz): remove this method
-    void reorderAtoms(const std::vector<unsigned int>& newOrder) {
-        std::vector<unsigned int> atomPositions;
```
Comment (b-scholz) [inline src/AstClause.h]: I am concerned with the latest changes whether the reordering is still working. The reason is that the reordering numbers were related to atoms in a body only.  However, in the new system, we may refer to atoms in schedules differently, ie., constraints/negations are counted as well. Is this correct?

Comment (azreika) [reply src/AstClause.h]: I took this into account when writing the replacement atom-reordering function; the function is semantically equivalent to filtering out all non-atoms, reordering the atoms based on the new order, and then putting back the non-atoms where they were before.

So, something like ‘a(x) :- b(x), !c(x), d(x)’ with new order ‘[1, 0]’ should result in ‘a(x) :- d(x), !c(x), b(x).’

 Good point, though - I will add a test to check for this.

**Comment by b-scholz** [comment]
> In a clause like:
> 
> `bad(Dog) :- bites(Dog, _Anybody).`
> 
> do we:
> 
> * Give a single-use warning for `_Anybody`, and treat it like a named variable, or
> * Not give a warning, and transform it into an unnamed variable in the back?\

I feel that not given a warning for this case, is the right solution. 

> 
> The system before this PR did not give a warning, but also did not treat it like an unnamed variable.
> 
> I am leaning towards the second, but then with a clause like:
> 
> `a(X) :- b(X, _X), c(_X).`
> 
> Do we want the `_X` to mean different things, or the same (i.e. bound to each other)?

I understand that _X is a variable with two occurrences whose prefix is _X. However, a warning should be issued 

> 
> This PR just gives a single-use warning for `_Anybody` and treats it like any other named variable.

I would suggest simulating the semantics of Prolog:
https://stackoverflow.com/questions/14238492/prolog-anonymous-variable

https://stackoverflow.com/questions/14238492/prolog-anonymous-variable

Perhaps

File: src/AstSemanticChecker.cpp [hunk 3]
```diff
@@ -660,12 +660,11 @@ void AstSemanticChecker::checkClause(ErrorReport& report, const AstProgram& prog
...
     // check for variables only occurring once
-    if (!clause.isGenerated()) {
-        for (const auto& cur : var_count) {
-            if (cur.second == 1 && cur.first[0] != '_') {
-                report.addWarning(
-                        "Variable " + cur.first + " only occurs once", var_pos[cur.first]->getSrcLoc());
-            }
+    for (const auto& cur : var_count) {
+        int numAppearances = cur.second;
+        std::string varName = cur.first;
```
Comment (mmcgr) [inline src/AstSemanticChecker.cpp]: This could be a ref, I think.

**Comment by azreika** [comment]
> > In a clause like:
> > `bad(Dog) :- bites(Dog, _Anybody).`
> > do we:
> > 
> > * Give a single-use warning for `_Anybody`, and treat it like a named variable, or
> > * Not give a warning, and transform it into an unnamed variable in the back?\
> 
> I feel that not given a warning for this case, is the right solution.
> 
> > The system before this PR did not give a warning, but also did not treat it like an unnamed variable.
> > I am leaning towards the second, but then with a clause like:
> > `a(X) :- b(X, _X), c(_X).`
> > Do we want the `_X` to mean different things, or the same (i.e. bound to each other)?
> 
> I understand that _X is a variable with two occurrences whose prefix is _. However, a warning should be issued.
> 
> > This PR just gives a single-use warning for `_Anybody` and treats it like any other named variable.
> 
> I would suggest simulating the semantics of Prolog:
> https://stackoverflow.com/questions/14238492/prolog-anonymous-variable

This has now been done. I've added a test to cover this warning, too:

`Warning: Variable _Anybody marked as singleton but occurs more than once in file ...`

**Comment by b-scholz** [comment]
> This has now been done. I've added a test to cover this warning, too:
> 
> `Warning: Variable _Anybody marked as singleton but occurs more than once in file ...`

Great - thanks!

File: tests/evaluation/independent_body2/independent_body2.err [hunk 4]
```diff
@@ -0,0 +1,6 @@
+Warning: Variable y only occurs once in file independent_body2.dl at line 55
```
Comment (mmcgr) [inline tests/evaluation/independent_body2/independent_body2.err]: y occurs twice in the line, let alone file. Could the error be clarified a little?

Comment (b-scholz) [reply tests/evaluation/independent_body2/independent_body2.err]: This is an artifact of the expansion. With disjunctions in the representation, we can take care of this.

**Pull request merged**